### DISCUSSION TITLE
Bluetooth pairing with my smartphone

### DISCUSSION THREAD

**Discussion by chrgweber** [pybricks/support#1554] [Q&A]
Hi to everybody,
I looked through you discussion, but I did not found solution.
My Goal: I want to control an EV3 by a smartphone transmitting numericmailbox telegrams by bluutooth (smartphone is the client)
The EV3 is the server and received the numericMailbox telegrams and controls two motors in the right way.
Unfornutatly, after pairing the establlished bluetooth connection is interrupted (appox. 5-10 Seconds after pairing).
I think that this is the reason why the ev3 cannot communicate with my smartphone.

Is there are to way to permanently be connected via bluetooth.

When I remove the SD-Card with Micropython and pybricks (Version 2.0.0) on it, and start the ev3 again, I can connect my smartphone the EV3 via bluetooth.

Here is my Code for the EV3.

Thanks in adavance

```
#!/usr/bin/env pybricks-micropython
from pybricks.hubs import EV3Brick
from pybricks.ev3devices import (Motor, TouchSensor, ColorSensor,
                                 InfraredSensor, UltrasonicSensor, GyroSensor)
from pybricks.parameters import Port, Stop, Direction, Button, Color
from pybricks.tools import wait, StopWatch, DataLog
from pybricks.robotics import DriveBase
from pybricks.media.ev3dev import SoundFile, ImageFile
from pybricks.messaging import BluetoothMailboxServer, NumericMailbox


# This program requires LEGO EV3 MicroPython v2.0 or higher.
# Click "Open user guide" on the EV3 extension tab for more information.


# Create your objects here.
ev3 = EV3Brick()


server = BluetoothMailboxServer()
mbox1 = NumericMailbox('MailB1', server)
mbox2 = NumericMailbox('MailB2', server)
# The server must be started before the client!
print('waiting for connection...')
ev3.screen.print('Hallo\n')
server.wait_for_connection()
print('connected!')
# In this program, the server waits for the client to send the first message
# and then sends a reply.
while(1):
    mbox1.wait()
    mbox2.wait()
    ev3.screen.print(mbox1.read() )
    ev3.screen.print(mbox2.read() )

```

**Comment by dlech** [comment]
What application are you using on the phone for an RFCOMM socket?

**Comment by chrgweber** [comment]
I am unsing a simple application written with MIT App Inventor, which seems to use RFCOMM sockets for communication. In a former project the bluetooth communication between app at the smartphone with MIT App inventor created works fine with the EV3. I programmed the bluetooth server at the EV3 with the EV3-Programmer-App from Lego. Now I want to switch to Micropython, but that is where the problem starts.
Attached you see the scretch-code from MIT-Appinventor which causes the error.
![Design_App](https://github.com/pybricks/support/assets/34003685/e7bedf54-c151-4be5-9b4c-27ef02cc0504)
![Scretch_Code](https://github.com/pybricks/support/assets/34003685/e513be26-67b1-4f72-8ff1-f9b9860bfc0e)

**Comment by dlech** [comment]
Could you swap the client and the server so that the EV3 is the client that initiates the connection?

**Reply by chrgweber** [reply]
HI Dlech,
thanks for your recommondation. 
The nice thing with the Smartphone to be the client and the EV3 to be the server is, that I can manually choose the correct EV3's-Mac-Adress in the app. In the ListPicker1 I can see all MACs and the names of the Bluetooth devices, which makes it ease to identify the correct MAC of my EV3.
I found out, when I used pybrick-micropython and do the pairing with my smartphone over the display of the EV3, the MAC-Adress of the EV3 in my ListPicker1 in the app is different to the MAC-Adress, which is listed, when I boot up the EV3 without the SD-Card containing the micropython environment. Why is this the case?
This might be the reason for my problem.
Kind regards
Christoph

**Reply by dlech** [reply]
I understand it is more convenient the way you have it, but if it is the difference between not working and working, it seems worth trying.

Without knowing how AppInventor implemented their code, it is hard to say what it is expecting that the official LEGO firmware provides and ev3dev doesn't seem to.

**Comment by chrgweber** [comment]
Sorry for my late replay. 
I switched the mit app into a BT Server and the LegoBrick into a BT Client, but it did not connect.
Again, after pairing the brick stays connected for 5-10 seconds and and it disconnects.

Here my code in the brick:
```
from pybricks.hubs import EV3Brick
from pybricks.ev3devices import (Motor, TouchSensor, ColorSensor,
                                 InfraredSensor, UltrasonicSensor, GyroSensor)
from pybricks.parameters import Port, Stop, Direction, Button, Color
from pybricks.tools import wait, StopWatch, DataLog
from pybricks.robotics import DriveBase
from pybricks.media.ev3dev import SoundFile, ImageFile
from pybricks.messaging import BluetoothMailboxClient, NumericMailbox
# This program requires LEGO EV3 MicroPython v2.0 or higher.
# Click "Open user guide" on the EV3 extension tab for more information.
# Create your objects here.
ev3 = EV3Brick()
SERVER = "XX:XX:XX:XX:X:XX"   # here I used my correct MAC-Address
client = BluetoothMailboxClient()
mbox1 = NumericMailbox("MailB1", client)
mbox2 = NumericMailbox("MailB2", client)

print("establishing connection...")
ev3.screen.print("Hallo")
client.connect(SERVER)
print("connected!")
# In this program, the client sends the first message and then waits for the
# server to reply.
while(1):
    mbox1.wait()
    ev3.screen.clear()
    speed_msg='Speed: '+str(mbox1)+'\n'
    ev3.screen.print(speed_msg)
    mbox2.wait()
    angle_msg='Speed: '+str(mbox1)+'\n'
    ev3.screen.print(angle_msg_msg)
    print(speed_msg)
    print(angle_msg)  
```

Here is the relevant code frim MIT App-Inventor:
![grafik](https://github.com/pybricks/support/assets/34003685/19d9370e-50cb-4471-a29b-73952fd55e4b)

I start first the BT Server and then the BT Client. The Client gives an error Message:
```
Starting: brickrun --directory="/home/robot/getting_started" "/home/robot/getting_started/main.py"
----------
establishing connection...
Traceback (most recent call last):
  File "/home/robot/getting_started/main.py", line 28, in <module>
  File "pybricks/messaging.py", line 321, in connect
  File "pybricks/messaging.py", line 318, in connect
  File "pybricks/bluetooth.py", line 168, in handle_request
OSError: [Errno 111] ECONNREFUSED
```

**Reply by chrgweber** [reply]
I got it 

First of all it was the wrong RFCOMM-Channel.
I also forgot to read the mailbox correctly.

I hacked it like:
```python
from pybricks.hubs import EV3Brick
from pybricks.ev3devices import (Motor, TouchSensor, ColorSensor,
                                 InfraredSensor, UltrasonicSensor, GyroSensor)
from pybricks.parameters import Port, Stop, Direction, Button, Color
from pybricks.tools import wait, StopWatch, DataLog
from pybricks.robotics import DriveBase
from pybricks.media.ev3dev import SoundFile, ImageFile
from pybricks.messaging import BluetoothMailboxClient, NumericMailbox
import pybricks.messaging as pm

pm.EV3_RFCOMM_CHANNEL=5

# This program requires LEGO EV3 MicroPython v2.0 or higher.
# Click "Open user guide" on the EV3 extension tab for more information.

# Create your objects here.
ev3 = EV3Brick()
#pybricks.messaging.EV3_RFCOMM_CHANNEL=5

SERVER = 'FC:19:99:DD:BD:8E'
#SERVER = "ev3dev"
ev3.screen.print("Hallo")
client = BluetoothMailboxClient()
wait(1000)
client.connect(SERVER)
wait(1000)
mbox1 = NumericMailbox("MailB1", client)
mbox2 = NumericMailbox("MailB2", client)

print("establishing connection...")


print("connected!")

# In this program, the client sends the first message and then waits for the
# server to reply.
while(1):
    mbox1.wait()
    ev3.screen.clear()
    speed_msg='Speed: '+str(mbox1.read())+'\n'
    ev3.screen.print(speed_msg)
    mbox2.wait()
    angle_msg='Speed: '+str(mbox2.read())+'\n'
    ev3.screen.print(angle_msg)
    print(speed_msg)
    print(angle_msg)
    wait(1000)
```

**Reply by dlech** [reply]
That great! Thanks for following up. 

Maybe it will work the way you wanted with the EV3 as the server now too with the RFCOMM channel fixed?
